README.md
Graphite-News
=============

There are two leading theories as to why Graphite-News was created: people
generally had nothing better to do than wonder which new types of information
were being stored in their Graphite/Carbon databases. Or the *second* theory; I
needed a little pet project to try out some programming languages, etc. I will
leave it up to you.

Functionality
-------------
Graphite-News keeps tabs on which new data sources appear in your Graphite
storage so that you can see what new functionality is being shipped inside
*your* application. This is then exposed in a easy and simple web interface.

Technology
----------
Build with (or exists despite the following things being used):
 * Go (for tailing, server side state, webserver)
    * [go-bindata](https://github.com/jteeuwen/go-bindata) (to keep 
      everything in 1 file for easy deployment)
    * [go-metrics](https://github.com/rcrowley/go-metrics) (of course 
      we keep tabs on our own metrics as well)
    * Tested on Go 1.3 and tip
    * Started to use [goxc](https://github.com/laher/goxc/) for cross platform
      compilation. (Work on a Mac but most graphite installs are on Ubuntu,
      bear with me, not my area of expertise. Tested once on Windows succesfully as well.)
 * jQuery
 * Bootstrap
 * [Travis.ci](https://travis-ci.org/ojilles/graphite-news) for running all the unit tests: [![Build Status](https://travis-ci.org/ojilles/graphite-news.svg?branch=master)](https://travis-ci.org/ojilles/graphite-news)

Installing
----------
Couple of options, but the easiest (assuming you have Go setup) is just running
the following:

    $ go get github.com/ojilles/graphite-news
    $ $GOPATH/bin/graphite-news -l $GOPATH/src/github.com/ojilles/graphite-news/creates.log

(That last command will get it up and running, but is obviously not how you
actually want to operate this piece of software.) If you have recently created new data sources, you will see those. If not, create a new one for testing purposes with:

    $ PORT=2003; SERVER=localhost; echo "local.random.diceroll 4 `date +%s`" | nc ${SERVER} ${PORT}

This should get you a `graphite-news` binary in `$GOPATH/bin`. Getting help gives you:

    $ graphite-news -h

Usage: graphite-news [-i sec] [-p port] [-s graphite url] [-r] [-d] -l logfile

  * d=false: If set, allow clients to delete recently created data sources
  * i=5000: Number of [ms] interval for Web UI's to update themselves. Clients only update their config every 5min
  * l=[]: One or more locations of the Carbon logfiles we need to tail. (F.ex. -l file1 -l file2 -l *.log)
  * p=2934: Port number the webserver will bind to (pick a free one please)
  * r=false: If set, report our own statistics every minute to a graphite host
  * rh="localhost:2003": Change the graphite host for pushing metrics towards
  * rp="graphite-news.metrics": Prepend all metric names with this string
  * s="http://localhost:8080": URL of the Graphite render API, no trailing slash. Apple rendezvous domains do not work (like http://machine.local, use IPs in that case)

The two important ones are the input (`-l` should point to the carbon logfile,
or whatever is storing the standard output of the carbon deamon) and the output
(`-s` the URL of the Graphite render API). Example:

    $ ~/graphite-news -s http://192.168.1.66:8080 -l /opt/graphite/log/launchctl-carbon*.stdout

The flag `-d` will enable a delete button in the UI. Once clicked, the
go-server will attempt to remove the data source from disk. This can come in
handy to keep your list of metrics clean if for some reason data sources were
created by accident or temporarily. I find this personally very handy (e.g. you
could now remove that `local.random.diceroll`), but the security risk is
entirely yours. Any web client never communicates the actual file name to
remove though which should limit the damage that can be done.  (Clients order
the server to remove the file associated with data source name so only those
can be removed, and even then only the ones Graphite News knows about (by
default limited to the last 100 created)).

If you are using Ansible, you can thank [ianunruh](https://github.com/ianunruh)
for providing an [Ansible
role](https://github.com/ianunruh/monitoring-ansible/tree/master/roles/graphite-news)
for Graphite News.

Usage of Webinterface
---------------------
After starting `graphite-news`, it will tail through your logfile in search for
notifications that new data sources have been created and store those in
memory. If you point a browser to it (default port 2934, configurable through
`-p`) you are presented with this UI:

![Screenshot](https://raw.githubusercontent.com/ojilles/graphite-news/master/docs/images/screenshot-1.png)

Walkthrough:

 * Top left: `Server Connection` shows you if the client (this web app) is able
   to talk to the server side component.
 * Top right:
   * Shows the number of data sources in the server, bit useless at the moment
   * A button with which you can freeze your client with (e.g. no more data
     retrieval from the server)
 * The rest of the UI is a long list of new data sources found with some
   information on them. Interactions:
   * Click on a line, will open that up and show you the graph for that metric
   * Click on any other line, the previous one will close and the one belonging
     to the new line opens
   * Click on that graph, and it gets closed
 * Do nothing: the client will keep polling the server for more data sources
   and once found will automatically refresh and put that onto the page.

As you can see, very simple interaction model!

Reporting statistics to Graphite
--------------------------------
The go server process collects metrics about itself as well. These can be
reported back to graphite by using the `-r` flag, which is by default off. Once
enabled it will report all metrics to your Grahite server, by default located
at `localhost:2003` but can be overridden by using the `-rh` flag. Metrics by
default will be reported under `graphite-news.metrics` at the top root, which
can be altered by using the `-rp` flag. Most likely, this will incite an
Inception-level experience to the operator once he views the Webinterface.

As an example, here is a screenshot from the `json` HTTP end-point (which is
used to ferry metrics from server to client-side browser):

![metrics-screenshot](https://raw.githubusercontent.com/ojilles/graphite-news/master/docs/images/metrics-screenshot.png)

Couple of notes:

 * `JSON` end-point is requested by a client every 5 seconds. As you can see in
   the screenshot, the `GET.json.one-minute` metric shows ~0.2 (as there was
   one client connected). Same with the
`cactiStyle(scaleToSeconds(nonNegativeDerivative(graphite-news.metrics.GET.json.count),1))`.
 * All timing metrics are expressed as nanoseconds. To get back to the more
   often used milliseconds in the web-domain, scale by 0.000001. For example:
   `sortByName(cactiStyle(scale(graphite-news.metrics.GET.{json}.*-percentile,0.000001),"si"))`
 * All requests that are not 4xx/5xx responses are getting tracked by their
   URL. General structure is to have the `HTTP Method` (e.g. `GET`), followed
   by the URL, followed by the various metrics.
 * All requests also get aggregated on the just the `HTTP Method`, followed by
   `__all_reqs` for easier top-level view. (F.ex.: `GET.__all_reqs.count`)
 * `tail.input_lines.count` and `tail.datasources.count` will let you know how
   many log-lines have been parsed and how many data sources have been found
   respectively. Here is an example graph to illustrate:

![lines-screenshot](https://raw.githubusercontent.com/ojilles/graphite-news/master/docs/images/lines-screenshot.png)


To Compile
-----------
By default (`go build` or `go install`) the binary created does include all
static assets (such as javascript, css) needed for proper functioning, but
those are **not pulled from the files on disk**, but rather from `bindata.go`.
Regenerating `bindata.go` can be done with the provided `build-dst.sh` script.
This will result with all fresh assets in the binary (able to just `scp` to
another machine for example).  Effectively, `build-dst.sh` only does:

`go-bindata -ignore=\\.swp$ assets/...`

Compiling across platforms is done via `goxc`, install via `go get
github.com/laher/goxc`, then run `goxc` in the working directory.


CHANGELOG

0.0.1-alpha	Work in progress, please do not use for anything, or install
		on any turing-complete device that is accessible by anyone 
		else but you. You have been warned.


main_test.go
package main

import (
	"fmt"
	"os"
	"testing"
)

func BenchmarkHello(b *testing.B) {
	for i := 0; i < b.N; i++ {
		ds := Datasource{Name: fmt.Sprintf("datasource %v", i)}
		addItemToState(ds)
	}
}

// All state is not reset
func TestDeleteFileNonExisting(t *testing.T) {
	file := "/slkjasd/wefoiwef/8c2f54b252f1a23b00768bf94bfa48db/"
	err := deleteFile(file)
	if err == true {
		t.Fatal(fmt.Sprintf("Deleting non existing file did not return error: %v", file))
	}
}
func TestDeleteFileNoName(t *testing.T) {
	file := ""
	err := deleteFile(file)
	if err == true {
		t.Fatal(fmt.Sprintf("Deleting file with no name does not return error: %v", file))
	}
}

func TestDeleteFile(t *testing.T) {
	file := "/tmp/graphite-news-test-file-safe-to-delete-8c2f54b252f1a23b00768bf94bfa48db"
	// remove the file if it exists, ignore errors, and defer one too
	// just to be sure everything is gone and clean after the tests
	os.Remove(file)
	defer os.Remove(file)

	// create the file
	filehandle, _ := os.Create(file)
	filehandle.Close()

	result := deleteFile(file)
	if result == false {
		t.Fatal(fmt.Sprintf("Deleting file returned error: %v", file))
	}

	// check filesystem to see file is gone
	_, err1 := os.Stat(file)
	if err1 == nil {
		t.Fatal(fmt.Sprintf("Tried deleting file, did not return error but still exists: %s", file))
	}
}

func TestGettingAsset(t *testing.T) {
	file := fmt.Sprintf("%vindex.html", staticAssetsURL)[1:]
	data, _ := Asset(file)
	if len(data) == 0 {
		t.Fatal(fmt.Sprintf("Could not load static asset: %v", file))
	}
}

func TestSingleDsIntoState(t *testing.T) {
	tmp := len(State.Vals)
	ds1 := Datasource{Name: "some name"}
	addItemToState(ds1)
	if (len(State.Vals) - tmp) != 1 {
		t.Fatal("Not able to add datasource to internal state")
	}

	// find it back
	ds2 := getDSbyName(ds1.Name)
	if ds1.Name != ds2.Name {
		t.Fatal("getDSbyName did not return DS that we know exists")
	}

	// now delete it
	if deleteDSbyName(ds1.Name) {
		fmt.Printf("%+v", State.Vals)
		// all went well
		if len(State.Vals) != 0 {
			t.Fatal("After removing only DS, state is not empty")
		}
		ds3 := getDSbyName(ds1.Name)
		if ds1.Name == ds3.Name {
			t.Fatal("getDSbyName did return DS that should not have been there")
		}
	} else {
		// fatal
		t.Fatal("Was not allowed to delete ds known to exist")
	}
}

func TestDuplicateEntriesIntoState(t *testing.T) {
	tmp := len(State.Vals)
	ds1 := Datasource{Name: "TestDuplicateEntriesIntoState same name"}
	ds2 := Datasource{Name: "TestDuplicateEntriesIntoState same name"}
	addItemToState(ds1)
	addItemToState(ds2)
	if (len(State.Vals) - tmp) > 1 {
		t.Fatal("Able to add more than one data source with the same name")
	}
}

func TestDontStoreMoreThan1k(t *testing.T) {
	// test that we dont keep more than maxState items in the state
	// and that when we go above it, we keep the last values (and not
	// the oldest ones)

	// reset internal state so that we know exactly what the last value
	// should be in the internal state
	State.Vals = nil
	const testString string = "TestDontStoreMoreThanMaxState, item: "

	ds := Datasource{Name: "tmp"}
	for i := 1; i < maxState+11; i++ {
		ds.Name = fmt.Sprintf("%v %v", testString, i)
		addItemToState(ds)
	}
	if len(State.Vals) > maxState {
		t.Fatal("Able to add more than maxState items into State.Vals")
	}
	if len(State.Vals) < maxState-10 {
		t.Fatal("Too few items in State (e.g. got reset somewhere in the middle?)")
	}

	lastItem := State.Vals[len(State.Vals)-1:][0]
	if lastItem.Name != ds.Name {
		t.Fatal(fmt.Sprintf("The expected last item (after adding more then maxState items) was [%v] but actually found [%v]!", ds.Name, lastItem.Name))
	}
}

func TestParsing(t *testing.T) {

	type testpair struct {
		incr int
		line string
	}

	// integer indicates with how much the count of internal state should go up
	// (could be 0, eg. not valid test input, or ds already captured)
	// Also, each line should have a non-nil Create_date
	var testCases = []testpair{
		{1, "launchctl-carbon.stdout:24/08/2014 17:59:53 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/df-Volumes-Recovery_HD/df_complex-free.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{0, "astt"},
		{0, "launchctl-carbon.stdout:24/08/2014 17:59:53 :: [creates] creating database file / (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "13/09/2014 23:10:56 :: [creates] creating database file /opt/graphite/storage/whisper/local/random/diceroll.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 17:59:54 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/df-Volumes-Recovery_HD/df_complex-reserved.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 17:59:54 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/df-Volumes-Recovery_HD/df_complex-used.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 20:59:54 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/df-Volumes-Media/df_complex-free.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 20:59:54 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/df-Volumes-Media/df_complex-reserved.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 20:59:54 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/df-Volumes-Media/df_complex-used.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 23:10:40 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/curl_xml-default/gauge-tvseries_watched-Babylon_5.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{1, "launchctl-carbon.stdout:24/08/2014 23:10:40 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/curl_xml-default/gauge-tvseries_total-Babylon_5.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
		{0, "launchctl-carbon.stdout:24/08/2014 23:10:40 :: [creates] creating database file /opt/graphite/storage/whisper/mac-mini_local/collectd/curl_xml-default/.wsp (archive=[(60, 525600), (600, 518400)] xff=None agg=None)"},
	}
	State.Vals = nil // start fresh
	prev_count := len(State.Vals)

	for _, test := range testCases {
		parseLine(test.line)

		if len(State.Vals) != prev_count+test.incr {
			t.Fatal(fmt.Sprintf("Parsed line, should have seen %v new entries, saw %v. Line: %v", test.incr, len(State.Vals)-prev_count, test))
		}
		prev_count = len(State.Vals)

		// Check date parseing
		last_ds := State.Vals[len(State.Vals)-1:][0]
		// Do any checking on the actual values for the data source (not complete yet)
		if last_ds.Create_date.IsZero() {
			t.Fatal(fmt.Sprintf("Data source has invalid Create_date: %+v", last_ds))
		}
		if len(last_ds.Name) < 1 {
			t.Fatal(fmt.Sprintf("Data source doesnt have proper Name: %+v", last_ds))
		}
	}
}


main.go
package main

// Implements a server that tails Graphite's logs
// finds out when new data sources become available
// and serves out a UI that lets you see/render those
// continuously.

// Written by J.A. Oldenbeuving / ojilles@gmail.com

import (
	"encoding/json"
	"flag"
	"fmt"
	"github.com/ActiveState/tail"
	"github.com/cespare/go-apachelog"
	"github.com/rcrowley/go-metrics"
	"log"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"
)

type (

	// Structure of a single data source. Anything in capitals will
	// get marshalled over to any connecting browser
	Datasource struct {
		Name        string    // bla.te.jfwoiejf.1MinuteRate, etc
		Create_date time.Time // Holds timestamp of when DS got created
		Params      string    // Holds things like retention schema's, etc
		filename    string    // /opt/graphite/whisper/etc
	}

	// Holds the state (all newly detected data sources)
	state struct {
		*sync.RWMutex // inherits locking methods
		Vals          []Datasource
	}

	// Holds all configuration items for main. Anything with a capital
	// will get marshalled towards any browsers connecting
	configuration struct {
		Version          string
		CompileTime      string
		JsonPullInterval int
		GraphiteURL      string
		ServerPort       int
		logfileLocation  loglocslice

		// If set, allow server to delete data sources that were encountered (not
		// random ones, only the ones in the State (e.g. the last maxState # of items)
		AllowDsDeletes bool

		// These are used for reporting Graphite-news' own
		// stats to a Graphite server. If that server is being monitored
		// by Graphite-news, you've recreated Inception!
		reporterGraphiteEnabled bool
		reporterGraphiteHost    string
		reporterGraphitePrep    string
	}

	// used for parsing Flags input params
	loglocslice []string
)

var (
	// Will hold version information from https://github.com/laher/goxc
	VERSION    string
	BUILD_DATE string

	// declare a globally scoped State variable, otherwise
	// the request handlers can't get to it. If there is a better
	// way to do this, plmk.
	State = &state{&sync.RWMutex{}, []Datasource{}}

	// Instantiate struct to hold our configuration
	C = configuration{JsonPullInterval: 5000}
)

const (
	// Maximum number of data sources to hold in memory before
	// pruning out as new ones come in
	maxState    int = 100
	myLogFormat     = log.Ldate | log.Ltime

	staticAssetsURL = "/assets/"
)

func (i *loglocslice) String() string {
	return fmt.Sprintf("%v", *i)
}

// Used to do Flag parsing of unbounded number of input logfiles
func (i *loglocslice) Set(value string) error {
	*i = AppendIfMissing(*i, value)
	return nil
}

func init() {
	if len(VERSION) == 0 {
		VERSION = "non-packaged"
		BUILD_DATE = "now"
	}
	C.Version = VERSION
	C.CompileTime = BUILD_DATE

	flag.IntVar(&C.JsonPullInterval, "i", 5000, "Number of [ms] interval for Web UI's to update themselves. Clients only update their config every 5min")
	flag.IntVar(&C.ServerPort, "p", 2934, "Port number the webserver will bind to (pick a free one please)")
	flag.StringVar(&C.GraphiteURL, "s", "http://localhost:8080", "URL of the Graphite render API, no trailing slash. Apple rendezvous domains do not work (like http://machine.local, use IPs in that case)")
	flag.Var(&C.logfileLocation, "l", "One or more locations of the Carbon logfiles we need to tail. (F.ex. -l file1 -l file2 -l *.log)")
	flag.BoolVar(&C.AllowDsDeletes, "d", false, "If set, allow clients to delete recently created data sources")
	flag.BoolVar(&C.reporterGraphiteEnabled, "r", false, "If set, report our own statistics every minute to a graphite host")
	flag.StringVar(&C.reporterGraphiteHost, "rh", "localhost:2003", "Change the graphite host for pushing metrics towards")
	flag.StringVar(&C.reporterGraphitePrep, "rp", "graphite-news.metrics", "Prepend all metric names with this string")

	flag.Usage = func() {
		fmt.Printf("Usage: graphite-news [-i sec] [-p port] [-s graphite url] [-r] [-d] -l logfile \n")
		fmt.Printf("Version: %v (Compiled at %v). Code over at: https://github.com/ojilles/graphite-news/\n\n", VERSION, BUILD_DATE)
		flag.PrintDefaults()
	}
}

func makeHandler(fn func(http.ResponseWriter, *http.Request)) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		// For metric names, escape any natural dots in the name,
		// then replace any URL slashes with a dot (seems same semantics to me)
		// and make sure we dont have any starting or ending with a dot.
		metricUrl := strings.Replace(r.URL.Path, ".", "_", -1)
		metricUrl = strings.Replace(metricUrl, "/", ".", -1)
		metricUrl = strings.Trim(metricUrl, ".")
		if len(metricUrl) == 0 {
			metricUrl = "index_html"
		}

		metricName := fmt.Sprintf("%s.%s", r.Method, metricUrl)
		aggMetricName := fmt.Sprintf("%s.__all_reqs", r.Method)
		m := metrics.GetOrRegisterTimer(metricName, metrics.DefaultRegistry)
		mGet := metrics.GetOrRegisterTimer(aggMetricName, metrics.DefaultRegistry)
		mGet.Time(func() {
			m.Time(func() {
				fn(w, r)
			})
		})
	}
}

func jsonHandler(w http.ResponseWriter, r *http.Request) {
	State.RLock() // grab a lock, but then don't forget to
	js, err := json.Marshal(State.Vals)
	State.RUnlock() // unlock it again once we're done

	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.Write(js)
}

func statsHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	metrics.WriteJSONOnce(metrics.DefaultRegistry, w)
}

func configHandler(w http.ResponseWriter, r *http.Request) {
	js, err := json.Marshal(C)
	if err != nil || len(js) < 1 {
		errorHandler(w, r, http.StatusNotFound)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	w.Write(js)
}

func deleteFile(dsFilename string) (err bool) {
	l := log.New(os.Stdout, "main	", myLogFormat)
	m := metrics.GetOrRegisterCounter("deletes", metrics.DefaultRegistry)

	if len(dsFilename) < 1 {
		l.Printf("deleteFile called, ignoring b/c/o unlikely filename: %v", dsFilename)
		return false
	}

	if _, err := os.Stat(dsFilename); os.IsNotExist(err) {
		l.Printf("deleteFile called but no such file: %s", dsFilename)
		return false
	}

	removeErr := os.Remove(dsFilename)
	if removeErr != nil {
		l.Printf("deleteFile called but failed os.Remove call: %s", dsFilename)
		return false
	}

	l.Printf("deleteFile called and succeeded: %s", dsFilename)
	m.Inc(1)
	return true
}

func deleteHandler(w http.ResponseWriter, r *http.Request) {
	l := log.New(os.Stdout, "main	", myLogFormat)

	if r.Method != "POST" {
		l.Printf("DELETE called, ignoring b/c was not a post: %v:%v\n", r.Method, r.URL)
		// Only allow POSTs to delete DS's
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	r.ParseForm()
	dsName := r.PostFormValue("datasourcename")
	ds := getDSbyName(dsName)
	Success := false

	if (len(ds.Name) > 0) && (len(ds.filename) > 0) {
		Success = deleteFile(ds.filename)
	}

	if Success == true {
		_ = deleteDSbyName(dsName)
		w.Write(nil)
	} else {
		http.Error(w, "", http.StatusInternalServerError)
	}

	l.Printf("DELETE called for '%v' (filename: %v) with result: '%v'",
		dsName, ds.filename, Success)
}

func frontpageHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "text/html")
	data, err := Asset("assets/index.html")
	if err != nil || len(data) == 0 || r.URL.String() != "/" {
		errorHandler(w, r, http.StatusNotFound)
		return
	}
	w.Write(data)
}

func staticHandler(w http.ResponseWriter, r *http.Request) {
	// URLs come in like "/assets/js/xxx", then get transformed to
	//                   "assets/static/js/xxx"
	// E.g. skip leading slash, inject static, where /assets/ is a constant
	filePath := fmt.Sprintf("%vstatic/%v",
		staticAssetsURL,
		r.RequestURI[len(staticAssetsURL):])[1:]
	tmp, err := Asset(filePath)
	if err != nil || len(tmp) == 0 {
		errorHandler(w, r, http.StatusNotFound)
		return
	}
	w.Write(tmp)
}

func faviconHandler(w http.ResponseWriter, r *http.Request) {
	data, err := Asset("assets/favicon.ico")
	if err != nil {
		errorHandler(w, r, http.StatusNotFound)
		return
	}
	w.Write(data)
}

func errorHandler(w http.ResponseWriter, r *http.Request, status int) {
	w.WriteHeader(status)
	if status == http.StatusNotFound {
		fmt.Fprint(w, "<b>404</b>: Not Found")
	}
}

func parseTime(s string) time.Time {
	// 24/08/2014 20:59:54
	var t, _ = time.Parse("02/01/2006 15:04:05", s)
	return t
}

func addItemToState(ds Datasource) {
	var foundDuplicate = false
	l := log.New(os.Stdout, "tail	", myLogFormat)
	m_ds := metrics.GetOrRegisterCounter("tail.datasources", metrics.DefaultRegistry)

	if strings.HasSuffix(ds.Name, ".") {
		return
	}

	// Find out if we already have one with the same name, if
	// so skip it.
	State.RLock()

	for _, item := range State.Vals {
		if !foundDuplicate && ds.Name == item.Name {
			foundDuplicate = true
		}
	}

	// Can not defer this one, as it would prevent any write lock below from
	// getting aqcuired succesfully (race cond.)
	State.RUnlock()

	if !foundDuplicate {
		// We're writing to shared datastructures, grab a Write-lock
		State.Lock()
		defer State.Unlock()
		State.Vals = append(State.Vals, ds)

		if len(State.Vals) > maxState {
			State.Vals = State.Vals[len(State.Vals)-maxState : len(State.Vals)]
		}

		l.Printf("New datasource: %+v (total: %v)", ds.Name, len(State.Vals))
		defer m_ds.Inc(1)
	}
}

func parseLine(line string) {
	m_lines := metrics.GetOrRegisterCounter("tail.input_lines", metrics.DefaultRegistry)
	var dataPath = regexp.MustCompile(`[a-zA-Z\:]*([0-9].*) :: \[creates\] creating database file (.*/whisper/(.*)\.wsp) (.*)`)
	m_lines.Inc(1)
	match := dataPath.FindStringSubmatch(line)
	if len(match) > 0 {
		ds := fmt.Sprintf("%s", strings.Replace(match[3], `/`, `.`, -1))
		tmp := Datasource{Name: ds, Create_date: parseTime(match[1]), Params: match[4], filename: match[2]}
		addItemToState(tmp)
	}

}

func deleteDSbyName(dsName string) bool {
	if len(getDSbyName(dsName).Name) == 0 {
		return false
	}

	State.Lock()
	defer State.Unlock()

	for i, ds_tmp := range State.Vals {
		if ds_tmp.Name == dsName {
			State.Vals = append(State.Vals[:i], State.Vals[i+1:]...)
			return true
		}
	}
	return false
}

func getDSbyName(dsName string) Datasource {
	State.RLock()
	defer State.RUnlock()

	for _, ds_tmp := range State.Vals {
		if ds_tmp.Name == dsName {
			return ds_tmp
		}
	}
	x := Datasource{}
	return x
}

func tailLogfile(c chan string, file string) {
	l := log.New(os.Stdout, "main	", myLogFormat)
	tc := tail.Config{Follow: true, ReOpen: true, MustExist: true}
	t, err := tail.TailFile(file, tc)
	if err == nil {
		l.Print(fmt.Sprintf("Tailing File:[%s]\n", file))
		for line := range t.Lines {
			parseLine(line.Text)
		}
	}
	c <- fmt.Sprintf("%s", err)
}

func tailLogfiles(c chan string) {
	var files []string

	// loop through the configured locations, and do filesystem
	// globbing, building up a new list. There is probably a special
	// place in POSIX-hell for me :-)
	for _, file := range C.logfileLocation {
		matches, _ := filepath.Glob(file)
		for _, match := range matches {
			files = AppendIfMissing(files, match)
		}
	}
	for _, file := range files {
		go tailLogfile(c, file)
	}
}

func AppendIfMissing(slice []string, i string) []string {
	for _, ele := range slice {
		if ele == i {
			return slice
		}
	}
	return append(slice, i)
}

func reportMetrics() {
	// Set up metrics registry
	if C.reporterGraphiteEnabled {
		addr, _ := net.ResolveTCPAddr("tcp", C.reporterGraphiteHost)
		go metrics.Graphite(metrics.DefaultRegistry, 60e9, C.reporterGraphitePrep, addr)
	}
}

func main() {
	error_channel := make(chan string)
	l := log.New(os.Stdout, "main	", myLogFormat)
	flag.Parse()

	// grab any remaining arguments and pretend they belong
	// to -l :-) (Also solves the -l * case for example)
	for _, argument := range flag.Args() {
		C.logfileLocation = AppendIfMissing(C.logfileLocation, argument)
	}

	// Set up web handlers in goroutines
	mux := http.NewServeMux()
	mux.HandleFunc("/json/", makeHandler(jsonHandler))
	mux.HandleFunc("/stats/", makeHandler(statsHandler))
	mux.HandleFunc("/config/", makeHandler(configHandler))
	mux.HandleFunc("/delete/", makeHandler(deleteHandler))

	// These are all handled by the compiled in Assets
	mux.HandleFunc("/", makeHandler(frontpageHandler))
	mux.HandleFunc("/favicon.ico", makeHandler(faviconHandler))
	mux.HandleFunc(staticAssetsURL, makeHandler(staticHandler))

	// Add the logging handler for Apache Common-ish log output
	loggingHandler := apachelog.NewHandler(mux, os.Stdout)
	server := &http.Server{
		Addr:    fmt.Sprintf(":%v", C.ServerPort),
		Handler: loggingHandler,
	}

	go server.ListenAndServe()
	go tailLogfiles(error_channel)
	go reportMetrics()

	l.Println("Graphite News -- Showing which new metrics are available since 2014")
	l.Println(fmt.Sprintf("Version: %v (Compiled at %v). Code over at: https://github.com/ojilles/graphite-news/", VERSION, BUILD_DATE))
	l.Println(fmt.Sprintf("Graphite News -- http://localhost:%v		:: Main User Interface", C.ServerPort))
	l.Println(fmt.Sprintf("Graphite News -- http://localhost:%v/config/	:: Internal configuration in JSON", C.ServerPort))
	l.Println(fmt.Sprintf("Graphite News -- http://localhost:%v/stats/	:: Internal Metrics in JSON", C.ServerPort))
	l.Println(fmt.Sprintf("Graphite News -- http://localhost:%v/json/	:: JSON dump of new graphite data sources", C.ServerPort))
	l.Println(fmt.Sprintf("Configuration: %+v", C))
	// Wait for errors to appear then shut down
	l.Println(<-error_channel)
}
